pkg/scheduler/framework/runtime/batch.go

```diff
@@ -0,0 +1,219 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package runtime
+
+import (
+	"context"
+	"time"
+
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/klog/v2"
+	fwk "k8s.io/kube-scheduler/framework"
+	"k8s.io/kubernetes/pkg/scheduler/apis/config"
+	"k8s.io/kubernetes/pkg/scheduler/framework"
+	"k8s.io/kubernetes/pkg/scheduler/metrics"
+)
+
+type PodSignatureFunc func(p *v1.Pod) string
+
+// Only needed until we connect signatures
+func noBatchSignatures(p *v1.Pod) string { return "" }
+
+// OpportunisticBatching caches results from filtering and scoring when possible to optimize
+// scheduling of common pods.
+type OpportunisticBatch struct {
+	profileName   string
+	state         *batchState
+	lastCycle     schedulingCycle
+	signatureFunc PodSignatureFunc
+	handle        fwk.Handle
+}
+
+type batchState struct {
+	signature    string
+	sortedNodes  framework.SortedScoredNodes
+	creationTime time.Time
+}
+
+type schedulingCycle struct {
+	cycleCount int64
+	chosenNode string
+	succeeded  bool
+}
+
+const (
+	maxBatchAge = 500 * time.Millisecond
+)
+
+// Provide a hint for the pod based on filtering an scoring results of previous cycles. Caching works only for consecutive pods
+// with the same signature that are scheduled in 1-pod-per-node manner (otherwise previous scores could not be reused).
+// It's assured by checking the top rated node is no longer feasible (meaning the previous pod was successfully scheduled and the
+// current one does not fit). Note that this call does not modify state.
+func (b *OpportunisticBatch) GetNodeHint(ctx context.Context, pod *v1.Pod, state fwk.CycleState, nodeInfos fwk.NodeInfoLister, cycleCount int64) (string, string) {
+	logger := klog.FromContext(ctx)
+
+	signature := b.signatureFunc(pod)
+
+	// If we don't have state that we can use, then return an empty hint.
+	if !b.batchStateCompatible(ctx, logger, pod, signature, cycleCount, state, nodeInfos) {
+		metrics.BatchUsageStats.WithLabelValues(metrics.BatchResultNoHint).Inc()
+		logger.V(3).Info("OpportunisticBatch no node hint available for pod",
+			"profile", b.profileName, "pod", pod.GetUID(), "signature", signature, "cycleCount", cycleCount)
+		return "", signature
+	}
+
+	// Otherwise, pop the head of the list in our state and return it as
+	// a hint. Also record it in our data to compare on storage.
+	metrics.BatchUsageStats.WithLabelValues(metrics.BatchResultHint).Inc()
+	hint := b.state.sortedNodes.Pop()
+	logger.V(3).Info("OpportunisticBatch provided node hint",
+		"profile", b.profileName, "pod", pod.GetUID(), "cycleCount", cycleCount, "hint", hint,
```

Comment:
Pass `klog.KObj(pod)` in the logs. UID is less meaningful
```suggestion
		"profile", b.profileName, "pod", klog.KObj(pod), "cycleCount", cycleCount, "hint", hint,
```